src/main/java/ws/wamp/jawampa/WampRouter.java

```diff
@@ -675,32 +701,70 @@ private void onMessageFromRegisteredChannel(WampRouterHandler handler, WampMessa
             long publicationId = IdGenerator.newRandomId(null); // Store that somewhere?
 
             // Get the subscriptions for this topic on the realm
-            Subscription subscription = handler.realm.subscriptionsByTopic.get(pub.topic);
-            if (subscription != null) {
-                for (WampRouterHandler receiver : subscription.subscribers) {
-                    if (receiver == handler) { // Potentially skip the publisher
-                        boolean skipPublisher = true;
-                        if (pub.options != null) {
-                            JsonNode excludeMeNode = pub.options.get("exclude_me");
-                            if (excludeMeNode != null) {
-                                skipPublisher = excludeMeNode.asBoolean(true);
-                            }
+            Subscription exactSubscription = handler.realm.subscriptionsByType.get(Subscription.Type.exact).get(pub.topic);
+            if (exactSubscription != null) {
+                publishEvent(handler, pub, publicationId, exactSubscription);
+            }
+
+            Map<String, Subscription> prefixSubscriptionMap = handler.realm.subscriptionsByType.get(Subscription.Type.prefix);
+            for (Subscription prefixSubscription : prefixSubscriptionMap.values()) {
```

Comment:
I'm not a big fan that we have to loop through all subscriptions pattern based subscriptions that were made on the router for each publication. Maybe subscriptions should be stored in some kind of tree structure where each uri component corresponds to a tree node? 

However this is an optimization question. Could be postponed until someone runs into problems with it.